src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -759,7 +756,7 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
             }
             else if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
             {
-                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
+                return AllocateArray<T>(length, pinned: true);
```

Comment:
```suggestion
                // Deferring to the initializing allocation API since elements
                // may contain references.
                return AllocateArray<T>(length, pinned: true);
```

If the above isn't appropriate, then I think it would make sense to just set `pinned = true;` and fall through.

Reply:
The whole `RuntimeHelpers.IsReferenceOrContainsReferences<T>()` path can be deleted. 

The VM is going to force zero-initialization for arrays with references here: https://github.com/dotnet/runtime/blob/a864ec70925181f00c38e15e4b1896f2584953c4/src/coreclr/vm/gchelpers.cpp#L217-L218

Reply:
I considered that but not sure about what `GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL` means for ref types. I assumed since the frontend protects this path it may go bad.

Reply:
`GC_ALLOC_ZEROING_OPTIONAL` is effectively ignored for ref types. It should be fine to set it unconditionally here. The VM code that I have linked above should handle it.

Reply:
> I assumed since the frontend protects this path it may go bad.

The frontend check was there to give us freedom to drop the support in the GC implementation if we find it convenient.

Reply:
Alright good to know. In that case, I simplified the paths in NativeAOT and the normal CoreCLR.